state-sync/aptos-data-client/src/latency_monitor.rs

```diff
@@ -0,0 +1,498 @@
+                let timestamp_now_usecs = self.get_timestamp_now_usecs();
+                if let Some(duration_from_propose_to_sync) =
+                    calculate_duration_from_proposal(block_timestamp_usecs, timestamp_now_usecs)
+                {
+                    metrics::observe_value_with_label(
+                        &metrics::SYNC_LATENCIES,
+                        metrics::PROPOSE_TO_SYNC_LATENCY_LABEL,
+                        duration_from_propose_to_sync.as_secs_f64(),
+                    );
+                }
+            }
+        }
+
+        // Update the advertised versions with those we still need to sync
+        *self.advertised_version_timestamps.write() = unsynced_advertised_versions;
+    }
+
+    /// Updates the advertised version timestamps by inserting any newly seen versions
+    /// into the map and garbage collecting any old versions.
+    fn update_advertised_version_timestamps(
+        &mut self,
+        highest_synced_version: u64,
+        highest_advertised_version: u64,
+    ) {
+        // Check if we're still catching up to the latest version
+        if !self.caught_up_to_latest {
+            if highest_synced_version + MAX_VERSION_LAG_TO_TOLERATE >= highest_advertised_version {
+                info!(
+                    (LogSchema::new(LogEntry::LatencyMonitor)
+                        .event(LogEvent::CaughtUpToLatest)
+                        .message(
+                            "We've caught up to the latest version! Starting the latency monitor."
+                        ))
+                );
+                self.caught_up_to_latest = true; // We've caught up
+            } else {
+                return; // We're still catching up, so we shouldn't update the advertised version timestamps
+            }
+        }
+
+        // If we're already synced with the highest advertised version, there's nothing to do
+        if highest_synced_version >= highest_advertised_version {
+            return;
+        }
+
+        // Get the current time and timestamp (note: we store both because
+        // there isn't a clean way of converting between them when relying
+        // on the time service).
+        let time_now_instant = self.time_service.now();
+        let timestamp_now_usecs = self.get_timestamp_now_usecs();
+
+        // Insert the newly seen version into the advertised version timestamps
+        self.advertised_version_timestamps.write().insert(
+            highest_advertised_version,
+            (time_now_instant, timestamp_now_usecs),
+        );
+
+        // If the map is too large, garbage collect the old versions
+        while self.advertised_version_timestamps.read().len() > MAX_NUM_TRACKED_VERSION_ENTRIES {
+            // Remove the lowest version from the map by popping the first
+            // item. This is possible because BTreeMaps are sorted by key.
+            self.advertised_version_timestamps.write().pop_first();
```

Comment:
Can we just hold on to the previous write lock a little longer for the GC? It seems like getting the write lock, read lock, then write lock again just adds lock contention.

Given the current loop, maybe a really unlucky thread can starve because the size increases each time it's waiting for the write lock.

Reply:
Aah, it's worth calling out that this is single threaded, i.e., there is only thread that runs the `latency_monitor` loop, so there shouldn't be any contention. 😄 In fact, I've removed the locks entirely. I originally added them so that the monitor could be cloneable (made it easier to write tests, etc.,) but I've worked around it now 😸